app/src/main/scala/org/alephium/explorer/persistence/schema/EventSchema.scala

```diff
@@ -0,0 +1,53 @@
+// Copyright 2018 The Alephium Authors
+// This file is part of the alephium project.
+//
+// The library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the library. If not, see <http://www.gnu.org/licenses/>.
+
+package org.alephium.explorer.persistence.schema
+
+import slick.jdbc.PostgresProfile.api._
+import slick.lifted.{Index, PrimaryKey, ProvenShape}
+
+import org.alephium.explorer.api.model.Address
+import org.alephium.explorer.persistence.model.EventEntity
+import org.alephium.explorer.persistence.schema.CustomJdbcTypes._
+import org.alephium.protocol.model.{BlockHash, TransactionId}
+import org.alephium.util.TimeStamp
+
+object EventSchema extends SchemaMainChain[EventEntity]("events") {
+
+  class Events(tag: Tag) extends Table[EventEntity](tag, name) {
+    def blockHash: Rep[BlockHash]          = column[BlockHash]("block_hash", O.SqlType("BYTEA"))
+    def txHash: Rep[TransactionId]         = column[TransactionId]("tx_hash", O.SqlType("BYTEA"))
+    def contractAddress: Rep[Address]      = column[Address]("contract_address")
+    def inputAddress: Rep[Option[Address]] = column[Option[Address]]("input_address")
+    def timestamp: Rep[TimeStamp]          = column[TimeStamp]("block_timestamp")
+    def eventIndex: Rep[Int]               = column[Int]("event_index")
+    def fields: Rep[Array[Byte]]           = column[Array[Byte]]("fields")
+
+    def * : ProvenShape[EventEntity] =
+      (blockHash, txHash, contractAddress, inputAddress, timestamp, eventIndex, fields)
+        .<>((EventEntity.apply _).tupled, EventEntity.unapply)
+
+    def pk: PrimaryKey = primaryKey("events_pk", (blockHash, txHash, contractAddress))
```

Comment:
We cannot use `(blockHash, txHash, contractAddress)` as the primary key because a contract can emit multiple events in one transaction.

Reply:
oh should we add `eventIndex` in the `pk` then?

Reply:
No, a contract can emit multiple events with the same `eventIndex` in one transaction. Events emitted by a contract have an offset, but it's a bit tricky to get the event offset. And because of reorg, for the same event, different full nodes may have different event offsets. I'm not sure if it's a good idea to add the event offset to the primary key.

Reply:
mmm then what would be the correct `PK`? maybe no primary key and use and auto increment? but we might have duplicate issue if we are not careful.

Reply:
I think duplicate is no problem, events can also be duplicated on full node because of reorg, users need to check if the event is from the mainchain.

Reply:
We could introduce a new field `event_order_in_block` for the index of an event inside a block, like how we handle tx orders inside a block. Then we could use `(block_hash, event_order_in_block)` as the primary key.

Reply:
It makes totally sense, I misunderstood `event_index` and treated it as `even_order_in_block`